rust/arrow/src/datatypes/decimal.rs

```diff
@@ -0,0 +1,926 @@
+                }
+
+                as_decimal.precision = precision;
+
+                Ok(as_decimal)
+            }
+
+            fn from_i64(
+                n: i64,
+                precision: usize,
+                scale: usize,
+            ) -> Result<Self, ParseDecimalError> {
+                let mut as_decimal = $name::new(n as $native_ty, precision, 0);
+
+                if as_decimal.scale != scale {
+                    as_decimal.rescale(scale)
+                }
+
+                as_decimal.precision = precision;
+
+                Ok(as_decimal)
+            }
+
+            fn from_f64(
+                n: f64,
+                precision: usize,
+                scale: usize,
+            ) -> Result<Self, ParseDecimalError> {
+                $name::parse(n.to_string().as_str(), precision, scale)
+            }
+
+            fn parse(
+                string: &str,
+                precision: usize,
+                scale: usize,
+            ) -> Result<Self, ParseDecimalError> {
+                let mut as_decimal = $name::from_str(string)?;
+
+                if as_decimal.scale != scale {
+                    as_decimal.rescale(scale)
+                }
+
+                as_decimal.precision = precision;
+
+                Ok(as_decimal)
+            }
+
+            fn to_le_bytes(&self) -> [u8; 16] {
+                self.digits.to_le_bytes()
+            }
+
+            fn to_be_bytes(&self) -> [u8; 16] {
+                self.digits.to_be_bytes()
+            }
+
+            fn to_byte_slice(&self) -> Vec<u8> {
+                self.digits.to_le_bytes().to_vec()
+            }
+
+            fn from_bytes_with_precision_scale(
+                bytes: &[u8],
+                precision: usize,
+                scale: usize,
+            ) -> $name {
+                let as_array = bytes.try_into();
+                match as_array {
+                    Ok(v) if bytes.len() == 16 => $name {
+                        digits: <$native_ty>::from_le_bytes(v),
+                        precision,
+                        scale,
+                    },
+                    Err(e) => panic!(
+                        "Unable to load Decimal from bytes slice ({}): {}",
+                        bytes.len(),
+                        e
+                    ),
+                    _ => panic!(
+                        "Unable to load Decimal from bytes slice with length {}",
+                        bytes.len()
+                    ),
+                }
+            }
+        }
+
+        impl ToPrimitive for $name {
+            fn to_isize(&self) -> Option<isize> {
+                unimplemented!("Unimplemented to_isize for {}", stringify!($name))
+            }
+
+            fn to_usize(&self) -> Option<usize> {
+                unimplemented!("Unimplemented to_usize for {}", stringify!($name))
+            }
+
+            fn to_i8(&self) -> Option<i8> {
+                Some(self.get_signed_lead_part() as i8)
```

Comment:
I feel like this is likely to `panic!` (at least in debug builds) if the lead part is greater than 128 / less than -128. Would it perhaps be better to check for overflow and return `None`?